### PR TITLE
*: make insert with calculated value behave the same as MySQL.

### LINKED ISSUE #4488: compability of insert with column
SQL:

```
create database mysqltest;
use mysqltest; 
create table t1 (c int);
insert into mysqltest.t1 set mysqltest.t1.c = '1'; -- Can't find column &{{} mysqltest t1 c} 
drop database mysqltest;
use test;
create table t1 (a int, b int); 
insert into t1 SET a=1, b=a+1;
```

TiDB:

```
mysql> create database mysqltest;
Query OK, 0 rows affected (0.08 sec)

mysql> use mysqltest;
Database changed

mysql> create table t1 (c int);
Query OK, 0 rows affected (0.08 sec)

mysql> insert into mysqltest.t1 set mysqltest.t1.c = '1';
ERROR 1105 (HY000): Can't find column &{{} mysqltest t1 c}

mysql> create table t1 (a int, b int);
Query OK, 0 rows affected (0.08 sec)

mysql> insert into t1 SET a=1, b=a+1;
ERROR 1054 (42S22): Unknown column '' in 'field list'
```

MySQL will suceed all above sql.

### PR DESCRIPTION
**Pull request by bobotu**
This PR try to fix most cases of #4488.
Now `insert` can correctly handle calculated value based on other columns, 
such as `insert into t1 set b=a+1` and `insert into t1 (b) value (a)`.
Last, as mentioned in issue, make the error message more readable.

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
Can the sql in #4482 be fixed after this commit when there is no default value for columns?

**Comment by bobotu** [comment]
@XuHuaiyu It can be fixed, but need a little change in `fillDefaultValues`. I'll file a new commit soon.

**Comment by XuHuaiyu** [comment]
PTAL @winkyao

File: executor/write_test.go [hunk 1]
```diff
@@ -1171,3 +1171,43 @@ func (s *testSuite) TestIssue4067(c *C) {
...
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t1, t2, t3")
+
+	tk.MustExec("create table t1(a int, b int)")
+	tk.MustExec("insert into t1 set a=1, b=a+1")
+	tk.MustQuery("select a, b from t1").Check(testkit.Rows("1 2"))
+
+	tk.MustExec("create table t2(a int default 100, b int)")
+	tk.MustExec("insert into t2 set b=a+1, a=1")
```
Comment (winkyao) [inline executor/write_test.go]: add test case: `insert into t2 set a=1, b=a+1;` In MySQL, It will insert `1 2`

File: executor/write_test.go [hunk 2]
```diff
@@ -1171,3 +1171,43 @@ func (s *testSuite) TestIssue4067(c *C) {
...
+	tk.MustExec("insert into t1 values (a+2)")
+	tk.MustExec("insert into t1 values (a)")
+	tk.MustQuery("select * from t1 order by a").Check(testkit.Rows("0", "2"))
+
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a bigint not null, b bigint not null)")
+	tk.MustExec("insert into t value(b + 1, a)")
+	tk.MustExec("insert into t set a = b + a, b = a + 1")
+	tk.MustExec("insert into t value(1000, a)")
+	tk.MustQuery("select * from t order by a").Check(testkit.Rows("0 1", "1 1", "1000 1000"))
```
Comment (winkyao) [inline executor/write_test.go]: add the case issued by winoros in #4482 , `create table` without null flags.

Comment (winkyao) [reply executor/write_test.go]: add test cases like:

```
mysql> create table t(a int default 1);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values (a);
Query OK, 1 row affected (0.00 sec)

mysql> select * from t;
+------+
| a    |
+------+
|    1 |
+------+
1 row in set (0.00 sec)
```

Comment (winoros) [reply executor/write_test.go]: Please add the conner cases like `JSON`, `enum`, `key with auto increment` and so on.

**Comment by shenli** [comment]
@winkyao @winoros PTAL

**Comment by winkyao** [comment]
@bobotu any update?

**Comment by bobotu** [comment]
@winkyao I have added a bundle of test case as mentioned above.

File: executor/write_test.go (hunk 2 above)
Comment (bobotu) [reply executor/write_test.go]: I have added these test cases at L1221-L1275.

File: executor/write.go [hunk 3]
```diff
@@ -886,15 +886,54 @@ func (e *InsertValues) getRows(cols []*table.Column, ignoreErr bool) (rows [][]t
...
+	var defaultValueCols []*table.Column
+	for i, c := range e.Table.Cols() {
+		var err error
+		if c.IsGenerated() {
+			continue
+		} else if mysql.HasAutoIncrementFlag(c.Flag) {
+			row[i] = table.GetZeroValue(c.ToInfo())
+		} else {
+			row[i], err = table.GetColDefaultValue(e.ctx, c.ToInfo())
+			if table.IsNoDefault(err) && mysql.HasNotNullFlag(c.Flag) {
```
Comment (winoros) [inline executor/write.go]: If fail to go into this branch, we should handle this error and return?

**Comment by shenli** [comment]
@winoros PTAL

File: executor/write.go [hunk 4]
```diff
@@ -886,15 +886,56 @@ func (e *InsertValues) getRows(cols []*table.Column, ignoreErr bool) (rows [][]t
...
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
 	}
-	return e.fillRowData(cols, vals, ignoreErr)
+
+	return e.checkRowData(cols, len(list), hasValue, row, ignoreErr)
+}
+
+func (e *InsertValues) fillDefaultValues(row []types.Datum, ignoreErr bool) error {
```
Comment (winkyao) [inline executor/write.go]: hasValue columns is filled by default values too?

File: plan/planbuilder.go [hunk 5]
```diff
@@ -801,8 +803,6 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
 			b.err = ErrBadGeneratedColumn.GenByArgs(assign.Column.Name.O, tableInfo.Name.O)
 			return nil
 		}
-		// Here we keep different behaviours with MySQL. MySQL allow set a = b, b = a and the result is NULL, NULL.
```
[the remark this answers is not in the document]
Comment (bobotu) [reply plan/planbuilder.go]: But after merge this PR, result of `set a = b, b = a` is `NULL, NULL`.

File: executor/write_test.go [hunk 6]
```diff
@@ -1171,3 +1171,106 @@ func (s *testSuite) TestIssue4067(c *C) {
 	tk.MustExec("delete from t1 where id in (select id from t2)")
 	tk.MustQuery("select * from t1").Check(nil)
 }
+
+func (s *testSuite) TestInsertCalculatedValue(c *C) {
```
Comment (XuHuaiyu) [inline executor/write_test.go]: add test cases for more generated columns.

File: executor/write.go (hunk 4 above)
Comment (bobotu) [reply executor/write.go]: `fillDefaultValues` invoked before eval actual insert expression. It fill a empty row with default value, then this row role as a eval context.

File: executor/write.go [hunk 7]
```diff
@@ -886,15 +886,56 @@ func (e *InsertValues) getRows(cols []*table.Column, ignoreErr bool) (rows [][]t
...
+	var defaultValueCols []*table.Column
+	for i, c := range e.Table.Cols() {
+		var err error
+		if c.IsGenerated() {
```
[the remark this answers is not in the document]
Comment (bobotu) [reply executor/write.go]: It use zero value for `not null` column even in `strict mode`. The column filled by `fillDefaultValues` will not mark in `hasValue`, so `checkRowData` will check nullability and insert correct value for column such as `auto_increment` based on `sql_mode`.

File: executor/write.go [hunk 8]
```diff
@@ -886,15 +886,56 @@ func (e *InsertValues) getRows(cols []*table.Column, ignoreErr bool) (rows [][]t
...
+	hasValue := make([]bool, len(e.Table.Cols()))
+	if err := e.fillDefaultValues(row, ignoreErr); err != nil {
+		return nil, errors.Trace(err)
+	}
+
 	for i, expr := range list {
-		val, err := expr.Eval(nil)
-		vals[i] = val
+		val, err := expr.Eval(row)
+		if err != nil {
```
Comment (winkyao) [inline executor/write.go]: use e.filterErr(err) here.

File: executor/write.go [hunk 9]
```diff
@@ -929,6 +970,11 @@ func (e *InsertValues) fillRowData(cols []*table.Column, vals []types.Datum, ign
 		row[offset] = v
 		hasValue[offset] = true
 	}
+
+	return e.checkRowData(cols, len(vals), hasValue, row, ignoreErr)
+}
+
+func (e *InsertValues) checkRowData(cols []*table.Column, valLen int, hasValue []bool, row []types.Datum, ignoreErr bool) ([]types.Datum, error) {
 	err := e.initDefaultValues(row, hasValue, ignoreErr)
```
Comment (winkyao) [inline executor/write.go]: why do initDefaultValues again?

File: plan/planbuilder.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (winkyao) [reply plan/planbuilder.go]: ok. got it.

[the remark this answers is not in the document]
Comment (winkyao) [reply plan/planbuilder.go]: ok, got it.

File: executor/write.go (hunk 9 above)
Comment (bobotu) [reply executor/write.go]: `fillDefaultValues` just ignore `sql_mode` and use `zero value` or `null` for columns, the result row just a `eval context` so `set a = 1, b = a` can eval correctly. But the actual value for no value column is influenced by `sql_mode` and other conditions, so I simply ask `initDefaultValues` to fill the correct value for no explicit value columns.

File: executor/write.go [hunk 10]
```diff
@@ -886,15 +886,56 @@ func (e *InsertValues) getRows(cols []*table.Column, ignoreErr bool) (rows [][]t
...
+			row[i] = table.GetZeroValue(c.ToInfo())
+		} else {
+			row[i], err = table.GetColDefaultValue(e.ctx, c.ToInfo())
+			if table.IsNoDefault(err) && mysql.HasNotNullFlag(c.Flag) {
+				row[i] = table.GetZeroValue(c.ToInfo())
+			} else if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		defaultValueCols = append(defaultValueCols, c)
```
Comment (winkyao) [inline executor/write.go]: I think defaultValueCols is not necessary, just use `e.Table.Cols()` in ` table.CastValues(e.ctx, row, e.Table.Cols(), ignoreErr)` ?

File: executor/write.go (hunk 9 above)
Comment (winkyao) [reply executor/write.go]: can we do these check in fillDefaultValues?

File: executor/write.go [hunk 11]
```diff
@@ -886,15 +886,60 @@ func (e *InsertValues) getRows(cols []*table.Column, ignoreErr bool) (rows [][]t
...
+	// So `insert t set b = a + 1` can use the correct value of a.
+	if err := e.fillDefaultValues(row, hasValue, ignoreErr); err != nil {
+		return nil, errors.Trace(err)
+	}
+
 	for i, expr := range list {
-		val, err := expr.Eval(nil)
-		vals[i] = val
+		val, err := expr.Eval(row)
+		err = e.filterErr(err, ignoreErr)
```
Comment (winkyao) [inline executor/write.go]: if err = e.filterErr(err, ignoreErr); err != nil {
}

File: executor/write.go [hunk 12]
```diff
@@ -886,15 +886,60 @@ func (e *InsertValues) getRows(cols []*table.Column, ignoreErr bool) (rows [][]t
...
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		val, err = table.CastValue(e.ctx, val, cols[i].ToInfo())
+		err = e.filterErr(err, ignoreErr)
```
Comment (winkyao) [inline executor/write.go]: ditto

**Comment by bobotu** [comment]
@winkyao PTAL

File: executor/write.go [hunk 13]
```diff
@@ -886,15 +886,58 @@ func (e *InsertValues) getRows(cols []*table.Column, ignoreErr bool) (rows [][]t
 }
 
 func (e *InsertValues) getRow(cols []*table.Column, list []expression.Expression, ignoreErr bool) ([]types.Datum, error) {
-	vals := make([]types.Datum, len(list))
+	// row is a eval context, so `insert t set b = 1, a = b + 1` can eval correctly.
```
Comment (XuHuaiyu) [inline executor/write.go]: I think this comment is not that explicit.
Can this comment and comments int lone 893 and 894 be removed,
and add a comment for this function.

File: executor/write.go [hunk 14]
```diff
@@ -886,15 +886,58 @@ func (e *InsertValues) getRows(cols []*table.Column, ignoreErr bool) (rows [][]t
...
+		}
+
+		offset := cols[i].Offset
+		row[offset], hasValue[offset] = val, true
+	}
+
+	return e.fillGenColData(cols, len(list), hasValue, row, ignoreErr)
+}
+
+// fillDefaultValues fill a row followed by these rules:
```
Comment (XuHuaiyu) [inline executor/write.go]: s/ fill/ fills

File: executor/write.go [hunk 15]
```diff
@@ -980,6 +1028,9 @@ func (e *InsertValues) initDefaultValues(row []types.Datum, hasValue []bool, ign
 			// Just leave generated column as null. It will be calculated later
```
Comment (XuHuaiyu) [inline executor/write.go]: Should we add a comment for initDefaultValues
to distinguish it with fillDefaultValues easily.

File: executor/write.go [hunk 16]
```diff
@@ -980,6 +1028,9 @@ func (e *InsertValues) initDefaultValues(row []types.Datum, hasValue []bool, ign
...
 			// but before we check whether the column can be null or not.
 			needDefaultValue = false
+			if !hasValue[i] {
```
Comment (XuHuaiyu) [inline executor/write.go]: any test case for this check?

Comment (bobotu) [reply executor/write.go]: Test L1279-L1281:
```SQL
create table t(a int auto_increment key, b int);
set SQL_MODE=NO_AUTO_VALUE_ON_ZERO;
insert into t (b) value (a+1);
```
If we don't reset `a` to `NULL`, `a` will be `0`.

File: executor/write.go [hunk 17]
```diff
@@ -643,7 +643,10 @@ type InsertValues struct {
 	batchRows    int64
 	lastInsertID uint64
 	ctx          context.Context
-	SelectExec   Executor
+
```
Comment (winkyao) [inline executor/write.go]: remove the blank line.

**Pull request merged**